arch/riscv/kvm/cove.c

```diff
@@ -980,14 +988,73 @@ int kvm_riscv_cove_vm_init(struct kvm *kvm)
 	return rc;
 }
 
+int kvm_riscv_cove_vm_single_step_init(struct kvm_vcpu *vcpu, unsigned long fdt_address,
+				unsigned long tap_addr)
+{
+	struct kvm *kvm = vcpu->kvm;
+	struct kvm_cove_tvm_context *tvmc;
+	unsigned long tvm_gid, target_vcpuid;
+	struct kvm_vcpu *target_vcpu;
+	void *nshmem = nacl_shmem();
+	struct kvm_cpu_context *cp = &vcpu->arch.guest_context;
+	int rc = 0, gpr_id;
+
+	tvmc = kzalloc(sizeof(*tvmc), GFP_KERNEL);
+	if (!tvmc)
+		return -ENOMEM;
+
+	for (gpr_id = 1; gpr_id < 32; gpr_id++) {
+		nacl_shmem_gpr_write_cove(nshmem, KVM_ARCH_GUEST_ZERO + gpr_id * sizeof(unsigned long),
+									((unsigned long *)cp)[gpr_id]);
+	}
+	kvm_arch_vcpu_load(vcpu, smp_processor_id());
```

Comment:
We set up the parameters required by the security monitor to create a TVM. These parameters are in: (1) NACL shared memory, (2) CSRs, (3) SBI ecall arguments (ie GPRs).

In the NACL memory we store the boot hart GPRs and CSRs.